### ISSUE TITLE
Add another point where middleware can attach so I can customize Inversify even more

### ISSUE THREAD

**Issue by rulai-hu** [inversify/InversifyJS#504]
**tl;dr** I'm using Inversify without TypeScript. I really like [constitute.js](https://github.com/justmoon/constitute)'s method of encoding types that will be injected since it hides so much implementation detail, so I wanted to make Inversify do the same via it's middleware, but that didn't work out. I hacked an less-than-ideal example together that could combine the best of both (by wrapping Inversify). [My implementation](https://gist.github.com/rulai-hu/e8943746f7c4f913c11d0a2e4a55609b) and [my unit tests](https://gist.github.com/rulai-hu/eab2dbeb5c20285b5457e78eb3fa44fb) (for Jest).

## Expected Behavior
In ES6 (without Babel decorators), I need to explicitly attach metadata via `Inversify.decorate` to a class in order to encode the dependencies that will be injected into that class. For me, this method is less than ideal because it 'leaks' the abstraction of the container to classes that have no business knowing that it exists. A static method is also pretty bad, but I gotta encode type information somewhere *shrug*. 
```
class A1 {}
class A {
  constructor(a1) {
    this.a1 = a1;
  }
}

inversify.decorate(inversify.injectable(), A);
inversify.decorate(inversify.injectable(), A1);
inversify.decorate(inversify.inject('A1Type'), A, 0);
Container.bind('AType').to(A);
Container.bind('A1Type').to(A1);
```
It's a bit cleaner with the helper:
```
register('A1Type')(A1);
register('AType', ['A1Type'])(A);
```
## Current Behavior
I want to achieve this:
```
class A1 {}
class A {
  static dependencies() { return ['A1Type']; }
  /* or even: 
  static dependencies() { return [A1]; }
  */
  constructor(a1) {
    this.a1 = a1;
  }
}
```
The ideal behaviour is before the planning phase, a call to `get` first traverses the tree, annotating dependencies where it's needed. I make an performance-motivated assumption (although I recognize there are lot's of gotchas with it) that subtrees where the root is already annotated do not need to be traversed.

## Possible Solution
In the course of implementing this, I tried using Inversify's `applyMiddleware` functionality to intercept requests and apply annotation to all nodes in the request tree. But the middleware runs 'too early', and an overridden `contextInterceptor` runs too late. The only way is to wrap `Container.get` and do the annotation before anything happens at all. Less than ideal.

With this method of type hinting, there's no need to require Inversify decorators, or any helpers for most use cases. 

## Context
Official support, (e.g adding another point to attach middleware) will make this less hacky to implement. Also, although this request is a bit more out-there, if `Inversify.decorate` could return `false` for failure instead of throwing, and `true` for success, then I wouldn't have to add another bit of metadata to ensure a class isn't decorated twice (an empty catch block is really gross).

And this is a long shot, but official support for this method of declaring dependencies would be amazing. 😄

**Comment by Dirrk** [comment]
Hey @rulai-hu this is a well thought out idea and you have great examples.  I was wondering what this would give you that the inversify-vanillajs-helpers doesn't?  I know you used it in your example, but I guess I am wondering why you went further with this if the helpers provided a work around.  Did the helpers not solve all of the problems or was it just out of wanting to do it?

And to clarify you are asking for 3 potential features right?
1. Extending middleware to provide a hook that will allow this type of overriding
2. making Inversify.decorate not throw - by adding in a param/new function that doesn't throw instead?
3. Officially supporting a static function on a class that annotates the dependencies


To give my opinion on this, I wrote a solution using the helpers:
```javascript
//...
const decorateForMe = function(c, type) {
	let decorators = [];
	if (c && c.inversify && typeof c.inversify === 'function') {
		decorators = c.inversify();
	}
	register(type, decorators)(c);
};

decorateForMe(A, 'A');
///...
```
The full code can be seen here: https://gist.github.com/Dirrk/19cfa0ff53ece4d139b4248036a022c0

I think having decorate not throw is a bad idea for everyone else, because they want to be informed that the decorate failed properly while running the code.  I also think that adding a static function on a class directly couples that class to the IoC which, to me seems wrong.  Obviously decorators couple as well but not as a method on the class itself.  I have not worked with the middle ware yet so I really have no help to offer there atm but I just wanted to give you my feedback before one of the maintainers looks at it.

**Comment by remojansen** [comment]
I think this is  possible and I also think it is a good idea :+1: I will investigate more this evening but right now I'm thinking the following:

In InversifyJS we use the decorators to generate metadata then we use [reflection_utils.ts](https://github.com/inversify/InversifyJS/blob/a92d13f9981275d1f3f7bfb4e722f4f7f0d11dd9/src/planning/reflection_utils.ts) to [read that metadata](https://github.com/inversify/InversifyJS/blob/ee8562ec8c0f05b536ae65fa10f233f54657e36d/src/planning/planner.ts#L168).

The reflection utils code is not very plugable and it is hard coded to use `Reflect.getMetadata` at the moment. 

We could refactor the reflection utils code into something that consumes a new interface named `MetadataReader`. 

InversifyJS will use a class by default: 

```ts
ReflectMetadataReader implements MetadataReader {
   // ...
}
```

We can allow users to provide their own implementations of `MetadataReader`:

```ts
constainer.applyCustomMetadataReader(myStaticMethodMetadataReader);
```

Inversify will continue to validate that there is no missing metadata as it does right now. The only thing that we are changing is the source of the metadata.

**Comment by rulai-hu** [comment]
@Dirrk @remojansen Thanks for responding!

@Dirrk I considered the Inversify helper route, but ultimately no matter how many levels of abstraction I use, `Inversify.decorate` is secretly pulling the strings in the background of the file. I also considered 'centralizing' all the Inversify-related boilerplate, e.g. in Container Modules, but imo spreading information about a single class to multiple files is a difficult design decision to make.

I wanted a solution where classes can exist independently of Inversify or its helpers, trading for that with a useless static method (without Inversify to interpret it).

@remojansen That would definitely work. Being able to intercepting the point before metadata is read would make a clean and direct solution feasible.

**Comment by Dirrk** [comment]
@rulai-hu, yeah I played around with it a bit more last night and you would basically have to continue to copy paste a bunch more stuff to use my example in a real application due to all the ways you can use inversify in an app so I can see how that becomes a problem.

**Comment by remojansen** [comment]
Hi guys, this is nearly implemented now. You can follow the process at https://github.com/inversify/InversifyJS/pull/507 I'm hoping to release this this weekend :smile:

**Comment by remojansen** [comment]
I just released inversify@3.2.0 which has support for this feature. I have added some unit tests:

https://github.com/inversify/InversifyJS/blob/master/test/features/metadata_reader.test.ts

And some info in the docs:

https://github.com/inversify/InversifyJS/blob/master/wiki/middleware.md#custom-metadata-reader

**Comment by remojansen** [comment]
@rulai-hu was this feature useful for you? Did you manage to resolve your use case?
